### DISCUSSION TITLE
About ComputerCraft turtles

### DISCUSSION THREAD

**Discussion by maraxin** [HexxyCraft/modpack#15] [General]
Well, CC turtles offers you a cheap craft for a lot of things to automate (farming, mod farms, mining etc.) while hexbound can do little to what these turtles can do (you can activate things with construct, which could be nice, but mostly it'll be useless). 
Of course consturcts are aesthetic, because this server is about hex casting, but as a normal player, I think that when it comes down to "easy recipe and more features vs harder recipe and less features", I would choose the first one.
So my opinion here is to disable turtles or (if people don't want to disable them all) disable some of them (farming, mining, crafting turtles). 

Any other opinions about CC turtles and hexbound constructs? Maybe I am wrong and they (constructs) can be useful at some point, where CC can't? I want to hear everyones opinion about this

**Comment by object-Object** [comment]
We could maybe tweak the turtle recipes instead of disabling them outright? Perhaps make them require something from Hex (charged amethyst? an impetus? a construct core/battery?).

(Moved this to a discussion, issues are more for specific concrete changes, this is a community discussion (and possibly a contentious one).)

**Reply by maraxin** [reply]
Interesting, as making recipe use hex could be nice and it would fix one of issues, but sadly, it won't really change anything, as CC turtles would still overshadow hexbound constructs.

I could suggest other choice, but it would be about making constructs more useful (attacking, using media etc.), as there is a nice potential in them.

**Reply by LlamaSageR** [reply]
While it would be nice to allow constructs to attack or use media, that is beyond the scope of this modpack to deliver. It'd be up to the mod developer to actually implement that (though I do recall hearing another construct being made which *can* attack).

**Reply by maraxin** [reply]
This is true, it would be beyond the scope of modpack... but here is the problem, the modpack is _about_ hex casting, and having something, that overshadows a part of it (even if its an addon) is weird for this modpack. 
It would be fully correct if this modpack wasn't about hex casting.

**Reply by LlamaSageR** [reply]
Writing response in a separate comment

**Comment by LlamaSageR** [comment]
Constructs and Turtles both share the basic niche of "Robot that interacts with the world based on how it is programmed," but the matter of interfacing them makes them distinct enough.

Turtles cannot respond to Hexes (Focal Ports are an option, but still not as powerful as directly interfacing with Hexes via Links and Broadcasts), and Constructs cannot respond to Computers (Computers are better at storing information, but the logistics of wireless communication is *complicated). Not to mention the unique behavior each robot has.

Having two different systems allows for more player agency in how they tackle solutions (Such as why Botania's Crafty Crate, Create's Mechanical Crafters, and Hexal's Craft Spell are all allowed). Not to mention how they can be combined to form really cool inventions! (Constructs can cast Scribe's Gambit, and so can theoretically directly interface with CC, and thus with other turtles. And Turtles can manipulate redstone to trigger Broadcasters or alter values in a Focal Port that is constantly being read by a Wisp)

While I do agree that Turtles are cheaper, the complexity of Lua is enough of a challenge to mitigate that issue. It's like how Hex Casting is balanced exclusively around it being hard to cast spells (and having an enormous pile of Amethyst, but that's not ever really an issue later on.) Still, changing the recipe is not a terrible idea.

And if a player just copy+pastes someone else's code into a Turtle, that's their decision.

**Reply by object-Object** [reply]
If we do want to change the recipe, I think phials might be a good option? Gates it behind Hex + enlightenment, so it's on somewhat more equal ground with constructs; but still keeps the recipe cheap, which is a major strength of turtles — speeding things up by making big turtle arrays.

**Reply by LlamaSageR** [reply]
Phials sounds like a good idea. Should explore other options as well though

**Reply by maraxin** [reply]
Yeah, consturcts do respond to some hexes, same for turtles responding to computer, which can be used to make some really complex things, but I was talking about something simple (like auto farming crops).

> Having two different systems allows for more player agency in how they tackle solutions (Such as why Botania's Crafty Crate, Create's Mechanical Crafters, and Hexal's Craft Spell are all allowed).

By the way, you forgot crafting turtle, which is by far the easiest one to automate (botania one is pain in general, mechanical isn't a real pain, but eh and hexal spell is 50/50)

> And if a player just copy+pastes someone else's code into a Turtle, that's their decision.

This is a main problem, copying the code almost destroys the entire complexity of Lua, as you can copy "auto crops farm turtle" or "3x3 mining turtle with mining all ores", while for consturcts you need to make a spell for it (construct spells are big in general).

By the way, I do like that turtles can somewhat access to hexes (focal port).

**Reply by LlamaSageR** [reply]
Let me address each of your comments:

> Yeah, consturcts do respond to some hexes, same for turtles responding to computer, which can be used to make some really complex things, but I was talking about something simple (like auto farming crops).

For one, Create completely trivializes crop farming even more than Turtles. Most simple automation can already be handled by Create. The main concept and point of Turtles and Constructs is for more precise, and advanced, automation.


> By the way, you forgot crafting turtle, which is by far the easiest one to automate (botania one is pain in general, mechanical isn't a real pain, but eh and hexal spell is 50/50)

The main point I was attempting to make was that alternate methods are allowed, as long as they do not entirely override each other. For instance:
- The Crafting Turtle is a very powerful tool for programmatically crafting items, with the ability to be sent signals and such to choose the recipe, but it is also very slow, not to mention the necessity of having to implement Lua (complaints about simply copying designs is addressed below).
- Create's Mechanical Crafters provide excellent in-place automated crafting, along with native compatibility with Create's logistics. However, to achieve adequate speeds, proper infrastructure needs to be built to support the RPM.
- Botania's Crafty Crate is an excellent single-block solution which is as fast as you can sort items into it. The only logistical backbone is the item sorting, which can be done via Vanilla methods, or by those provided by other mods.
- Hexal's Craft allows for instantaneous crafting via interfacing with the Mote Nexus, even allowing multiple different crafts within a single tick. It has the highest theoretical output, but the actual logistics of managing the stack and *when* to craft is the primary challenge, not to mention the challenges of actually setting it up. 

It's about player choice. Some solutions might be easier to work with, but they all have their own strengths and weaknesses. That's why I chose Crafting as an example. In the case of Constructs vs Turtles, they each have their abilities and limitations. 

> This is a main problem, copying the code almost destroys the entire complexity of Lua, as you can copy "auto crops farm turtle" or "3x3 mining turtle with mining all ores", while for consturcts you need to make a spell for it (construct spells are big in general).

Everything can be copied. There are schematics for Create and Botania, you can copy code for CC, and you can simply copy down patterns for Hex Casting by hand from the Forums (not to mention the possibility of using the Focal Port to literally copy-paste it!). The point of the modpack is to provide opportunities for technical shenanigans and contraptions. If the player chooses to ignore them, that is their choice.

> By the way, I do like that turtles can somewhat access to hexes (focal port).

While CC can trivialize some aspects of Hex design, it opens up so many other avenues. The more intercompatibility there is, the more possibilities there are. That's why I don't want to disable Turtles. 

(And yes, I like it too)

**Reply by maraxin** [reply]
> The main point I was attempting to make was that alternate methods are allowed, as long as they do not entirely override each other.

Override? lets see... I'll compare every auto crafting thing you said here with crafting turtle

> Create's Mechanical Crafters provide excellent in-place automated crafting, along with native compatibility with Create's logistics. However, to achieve adequate speeds, proper infrastructure needs to be built to support the RPM.

Fully overshadowed by CC turtles, painful to set up and automate, it does go with native compatibility, but crafting turtle can sort of make it as well. The only use is to craft special create stuff with mechanical crafter, otherwise there are lots of better variants of autocrafting.

> Botania's Crafty Crate is an excellent single-block solution which is as fast as you can sort items into it. The only logistical backbone is the item sorting, which can be done via Vanilla methods, or by those provided by other mods.

Barely anyone actually remembers this, the only logical reason is that it's not useful. Requires end game recipes from botania in order to set it up + its still painful to do it, crafting turtle wins here as well (+ not everyone actually knows how to set it up lmao).

> Hexal's Craft allows for instantaneous crafting via interfacing with the Mote Nexus, even allowing multiple different crafts within a single tick. It has the highest theoretical output, but the actual logistics of managing the stack and when to craft is the primary challenge, not to mention the challenges of actually setting it up.

Ah, it is clearly fast, as it can do lots of crafts in 1 tick, this is really nice & can be automated, as well as its one of main mods for this server, but requires mote nexus, which is not that hard to obtain, but requires shulkers and access to the end, which CC turtles don't require you to have so (also, with limited world border it might be even more painful to make it), also it barely has any compatibility with other mods (even though it is there) + requires lots of space to set up properly, while crafting turtle requires fuel and 1 block (which is turtle by itself). Sort of overshadowed by crafting turtle, but loses when it comes down to speed.

So what we see in the end? 2/3 auto crafts were overshadowed by CC, while 3rd (hexal) autocraft will be probably used more often, because this is hex casting server.

> Everything can be copied. There are schematics for Create and Botania, you can copy code for CC, and you can simply copy down patterns for Hex Casting by hand from the Forums. The point of the modpack is to provide opportunities for technical shenanigans and contraptions. If the player chooses to ignore them, that is their choice.

Yes, everything can be copied, create and botania schematics, as well as hex casting code from forums, but here's the catch... Copying code in CC requires a few commands, which shouldn't take more, than 1 minute to set up correctly (from my switchcraft 3 expirience, you can do lots of stuff without even knowing LUA at all, as I barely know it by myself).

> (not to mention the possibility of using the Focal Port to literally copy-paste it!)

Yup, soon with CC you'll be able to copy full hex casting code, but for now its not really good at it, but it still requires computers to actually copy.

> While CC can trivialize some aspects of Hex design, it opens up so many other avenues. The more intercompatibility there is, the more possibilities there are. That's why I don't want to disable Turtles.

I asked to disable some of them, because they're overshadowing other things (such as farming turtle overshadowing constructs harvest spell, or crafting turtle overshadowing almost every other way of crafting). Other way is to make turtles have end game recipe from botania, create and hexbound to make it fair.

**Reply by LlamaSageR** [reply]
In the interests of making this more legible, I will be breaking this post into several replies.

> Override? lets see... I'll compare every auto crafting thing you said here with crafting turtle

And here is my response.

> Fully overshadowed by CC turtles, painful to set up and automate, it does go with native compatibility, but crafting turtle can sort of make it as well. The only use is to craft special create stuff with mechanical crafter, otherwise there are lots of better variants of autocrafting.

Let me unpack this first:
- Painful to set up and automate. 

"Painful" is a subjective aspect. For you it might be challenging, but I am certain coding a Turtle would be similarly, if not more "painful" to others. Create provides all the tools needed to properly sort items for usage in a Mechanical Crafter, tools which are well documented in-game. A single mechanical arm with a properly organized belt can automate the entire thing single-handedly. This is not even considering if the player does not have coding experience already. Mechanical Crafters being in-world as opposed to magical black boxes makes them far more "reliable" to set up, since errors will manifest directly in the world as opposed to a typo killing the turtle in 3 hours. 

- It does go with native compatibility, but crafting turtle can sort of make it as well.

Many aspects of these technical mods can be approximated by others. Botania, Create, and CC can all automate basically every Vanilla aspect just on their own. The difference is how they go about doing that automation. Trying replace Mechanical Crafters with Crafting Turtles is just overcomplicating a simple process, assuming you have your belts properly organized.

- Only use is to craft special create stuff with mechanical crafter

That the static in-world crafting that can be ran indefinitely with a simple fuel source, faster crafting (allegedly, from what the others have said) than the Turtle, and an intuitive system that makes designs simple and fun to execute.

- Otherwise there are lots of better variants of autocrafting

Aren't you arguing that the only good method is using Turtles? Regardless, the other options aren't objectively "better," they're "different." Mechanical Crafters might be intuitive to design for, with the highly modular system Create provides, but it requires a dedicated system of infrastructure to provide the SU for high speeds, and it takes a large amount of space. Crafty Turtles are smaller, but they are slower crafters and more complicated to design for due to their more generalist nature, not to mention the logistics of feeding items to refuel them at set intervals.

**Reply by LlamaSageR** [reply]
> Barely anyone actually remembers this, the only logical reason is that it's not useful. Requires end game recipes from botania in order to set it up + its still painful to do it, crafting turtle wins here as well (+ not everyone actually knows how to set it up lmao).

And again, a breakdown.

- Barely anyone actually remembers this, the only logical reason is that it's not useful.

A false assumption. The Crafty Crate can be useful, the issue is that most packs that include Botania tend to be kitchen-sink packs, which include AE2, the ultimate form of logistics and auto-crafting. AE2 invalidates nearly every other method auto-crafting. The reason why other auto-crafters are remembered is that they are used in earlier in more progression themed packs or Expert packs. As Botania does not contribute to the more technological theme these packs tend to follow, it is often excluded and overshadowed, especially since it is more centered around Vanilla in terms of balance.

- Requires end game recipes from botania in order to set it up

All it requires is Elvin Portal access, which is post-Nether similarly to Create's Mechanical Crafters, and before Hexal's End-gated Mote Nexus. End-game Botania would be post Gaia Guardian. Regardless, I will concede that Turtles could be tweaked to be further along in progression.

-  its still painful to do it, crafting turtle wins here as well

And again you bring the subjective term "painful." Coding an entire Turtle I would wager would be "painful" as well. It is more challenging to manage the logistics via just Botania I will admit, but that is just one of Botania's weaknesses. Create can be used here to compensate for that weakness (or Hexbound's Constructs), or a cleverly designed Vanilla solution can be used. The Crafty Crate requires no fuel or energy, and crafts as fast as you can shove items into it, making it potentially faster than Create's system without the need for a dedicated power generator. As above, programming an entire Turtle from Lua is its own ordeal.

- (+ not everyone actually knows how to set it up lmao).

And not everyone knows how to use Create or Computer Craft. The Lexica Botania describes what is needed well enough, it is up to the Player at that point to actually implement it.

**Reply by LlamaSageR** [reply]
> Ah, it is clearly fast, as it can do lots of crafts in 1 tick, this is really nice & can be automated, as well as its one of main mods for this server, but requires mote nexus, which is not that hard to obtain, but requires shulkers and access to the end, which CC turtles don't require you to have so (also, with limited world border it might be even more painful to make it), also it barely has any compatibility with other mods (even though it is there) + requires lots of space to set up properly, while crafting turtle requires fuel and 1 block (which is turtle by itself). Sort of overshadowed by crafting turtle, but loses when it comes down to speed.

And again.

- Ah, it is clearly fast, as it can do lots of crafts in 1 tick, this is really nice & can be automated, as well as its one of main mods for this server

Agreed, the speed and flexibility of Hex Casting makes it very nice to use. 

- but requires mote nexus, which is not that hard to obtain, but requires shulkers and access to the end, which CC turtles don't require you to have so

A one-time investment which comes with the best singular source of storage in the modpack. Hexal's Craft is more akin to AE2's autocrafting mechanic, though far more restricted. CC Turtles are not better here simply because they are cheaper.

- (also, with limited world border it might be even more painful to make it)

It shouldn't be. If Shulkers are too limited, then a fix will be implemented for that. Regardless, it is a one-time investment for the most part.

- also it barely has any compatibility with other mods (even though it is there)

Not direct compatibility, agreed. But it is still highly compatible *indirectly*. The ability to wirelessly transmit items and automatically craft with them is a tool that cannot be underestimated, especially with the instantaneous aspect of it. Not to mention the plethora of tools Hex Casting proper provides and the interactions with Ducky's Focal Port.

- requires lots of space to set up properly, while crafting turtle requires fuel and 1 block (which is turtle by itself)

In line with the logistics Create's Mechanical Crafters require. Botania's Crafty Crate and the Crafty Turtles are far smaller in footprint, which is their advantage. Botania is static and requires intelligently designed logistics, Turtles are generalist and flexible, but require forethought and planning alongside coding and logistics to implement their behaviour (logistics can be ignored at the expense of more complicated code, a valid tradeoff).

- Sort of overshadowed by crafting turtle, but loses when it comes down to speed.

See, that's the thing you don't see with the other solutions. They all have different strengths and weaknesses, which lead to different use cases for each one. The Turtle is far more portable and flexible than the rest, but a dedicated setup is superior in speed and simplicity (not to mention how more accessible logistics systems are to stationary systems). Hexal's Craft is directly connected to the densest item storage in almost all of Modded Minecraft, the space required for it is simply the Spell Circle to implemented. As a result, it can be used in remote locations in nodes to craft dedicated items and funnel them back into the network based on demand. More flexible and programmatic than Create and Botania, but it suffers from the complexity of coding all the features. Not to mention the only method of importing items is via Spells, which is more inconvenient than Belts or Hoppers.

- So what we see in the end? 2/3 auto crafts were overshadowed by CC, while 3rd (hexal) autocraft will be probably used more often, because this is hex casting server.

What I anticipate to happen is that the Mechanical Crafter will be used the most (easily set up and simple to understand). The Craft Spell will be used for the novelty and challenge, Botania will likely be overshadowed by the ease of use of Create but still have a valued niche for the more industrially inclined (or those who want a challenge), and CC Turtles will be ignored due to the presence of easier options. It might see some use early, due to the cheap recipe, but the Nether can be reached in an hour, so players would simply wait for autocrafting until Mechanical Crafters can be obtained (not to mention that they need to develop their base first, which is typically post-nether).

But the point is that Players can use these options, and that they have some purpose. Different Players have different desires and needs, and so different Auto-Crafters are ideal to them.

**Reply by LlamaSageR** [reply]
And this will be the last section.

> Yes, everything can be copied, create and botania schematics, as well as hex casting code from forums, but here's the catch... Copying code in CC requires a few commands, which shouldn't take more, than 1 minute to set up correctly (from my switchcraft 3 expirience, you can do lots of stuff without even knowing LUA at all, as I barely know it by myself).

Counterpoint: Schematicannon. It can literally print it out for you. That, or some client-side mod like Litematica that guides you through the entire build. Not to mention that if you don't know LUA, you cannot modify the system when needed, as opposed to the simplicity of Mechanical Crafters or the in-depth experience most people on the server have with Hex Casting. 

> Yup, soon with CC you'll be able to copy full hex casting code, but for now its not really good at it, but it still requires computers to actually copy.

Valid point. Though someone will successfully implement it eventually, if someone has not already.

> I asked to disable some of them, because they're overshadowing other things (such as farming turtle overshadowing constructs harvest spell, or crafting turtle overshadowing almost every other way of crafting). Other way is to make turtles have end game recipe from botania, create and hexbound to make it fair.

The Farming Turtle is infinitely overshadowed by Create already and I've already addressed how the Crafting Turtle is not as powerful as believed. Not to mention that I doubt you can disable select turtles, since they can simply equip the items and perform the tasks, though that's besides the point. Making the Turtles more expensive is something we discussed before, and would be willing to elaborate on. A large part of the power and interest of Turtles is their mass numbers and cheap recipe, so it can't be too expensive.

**Reply by walksanatora** [reply]
~~we should remove HTTP support from CC just to make importing from the web harder (for everyone else, I have my own archive format allready)~~
(and I allready have implemented a way to load .hexpattern files directly into MC using Focal Ports, but that is something you can search through with my github)
